File: lambda/python/assets_handler.py
import json
import os
import boto3
import structlog
from typing import Dict, Any
from botocore.exceptions import ClientError

# Initialize clients
s3 = boto3.client('s3')
dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table(os.environ['ASSETS_TABLE_NAME'])
logger = structlog.get_logger()

def handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lambda handler for assets API endpoints
    
    Supports:
    - POST /assets - Upload asset
    - GET /assets/{id} - Get asset metadata
    - GET /assets/{id}/download - Get signed URL for download
    """
    try:
        logger.info("Asset request received", path=event.get('path'), method=event.get('httpMethod'))
        
        method = event.get('httpMethod')
        path = event.get('path', '')
        path_params = event.get('pathParameters') or {}
        body = json.loads(event.get('body', '{}'))
        
        asset_id = path_params.get('id')
        action = path_params.get('action', '')
        
        if method == 'POST':
            return upload_asset(body)
        elif method == 'GET' and asset_id:
            if action == 'download':
                return get_download_url(asset_id)
            else:
                return get_asset(asset_id)
        else:
            return {
                'statusCode': 405,
                'body': json.dumps({'error': 'Method not allowed'})
            }
            
    except json.JSONDecodeError:
        logger.error("Invalid JSON in request body")
        return {
            'statusCode': 400,
            'body': json.dumps({'error': 'Invalid JSON'})
        }
    except ClientError as e:
        logger.error("AWS service error", error=str(e))
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Internal server error'})
        }
    except Exception as e:
        logger.error("Unexpected error", error=str(e))
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Internal server error'})
        }


def upload_asset(body: Dict[str, Any]) -> Dict[str, Any]:
    """Create asset metadata in DynamoDB and return upload URL"""
    asset_id = body.get('assetId')
    file_name = body.get('fileName')
    content_type = body.get('contentType', 'application/octet-stream')
    
    if not asset_id or not file_name:
        return {
            'statusCode': 400,
            'body': json.dumps({'error': 'assetId and fileName are required'})
        }
    
    logger.info("Creating asset metadata", asset_id=asset_id, file_name=file_name)
    
    try:
        import time
        import uuid
        
        bucket_name = os.environ['ASSETS_BUCKET_NAME']
        s3_key = f"assets/{asset_id}/{file_name}"
        
        # Check if asset already exists (idempotency)
        existing = table.get_item(Key={'assetId': asset_id})
        if 'Item' in existing:
            logger.info("Asset already exists", asset_id=asset_id)
            return {
                'statusCode': 200,
                'body': json.dumps({'message': 'Asset already exists', 'asset': existing['Item']})
            }
        
        # Store metadata in DynamoDB
        item = {
            'assetId': asset_id,
            'fileName': file_name,
            'contentType': content_type,
            's3Key': s3_key,
            's3Bucket': bucket_name,
            'status': 'pending',
            'createdAt': int(time.time()),
            'ttl': int(time.time()) + (30 * 24 * 60 * 60)  # TTL: 30 days
        }
        
        response = table.put_item(
            Item=item,
            ConditionExpression='attribute_not_exists(assetId)'
        )
        
        # Generate presigned URL for upload
        upload_url = s3.generate_presigned_url(
            'put_object',
            Params={
                'Bucket': bucket_name,
                'Key': s3_key,
                'ContentType': content_type
            },
            ExpiresIn=3600  # 1 hour
        )
        
        return {
            'statusCode': 201,
            'body': json.dumps({
                'message': 'Asset metadata created',
                'asset': item,
                'uploadUrl': upload_url
            })
        }
        
    except ClientError as e:
        if e.response['Error']['Code'] == 'ConditionalCheckFailedException':
            existing = table.get_item(Key={'assetId': asset_id})
            if 'Item' in existing:
                return {
                    'statusCode': 200,
                    'body': json.dumps({'message': 'Asset already exists', 'asset': existing['Item']})
                }
        raise


def get_asset(asset_id: str) -> Dict[str, Any]:
    """Get asset metadata"""
    logger.info("Getting asset", asset_id=asset_id)
    
    try:
        response = table.get_item(Key={'assetId': asset_id})
        
        if 'Item' not in response:
            return {
                'statusCode': 404,
                'body': json.dumps({'error': 'Asset not found'})
            }
        
        return {
            'statusCode': 200,
            'body': json.dumps(response['Item'])
        }
        
    except ClientError as e:
        logger.error("DynamoDB error", error=str(e))
        raise


def get_download_url(asset_id: str) -> Dict[str, Any]:
    """Get presigned URL for downloading asset"""
    logger.info("Getting download URL", asset_id=asset_id)
    
    try:
        response = table.get_item(Key={'assetId': asset_id})
        
        if 'Item' not in response:
            return {
                'statusCode': 404,
                'body': json.dumps({'error': 'Asset not found'})
            }
        
        item = response['Item']
        download_url = s3.generate_presigned_url(
            'get_object',
            Params={
                'Bucket': item['s3Bucket'],
                'Key': item['s3Key']
            },
            ExpiresIn=3600  # 1 hour
        )
        
        return {
            'statusCode': 200,
            'body': json.dumps({
                'downloadUrl': download_url,
                'fileName': item['fileName'],
                'contentType': item.get('contentType')
            })
        }
        
    except ClientError as e:
        logger.error("S3/DynamoDB error", error=str(e))
        raise

File: lambda/typescript/src/assets-handler.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand, PutCommand } from '@aws-sdk/lib-dynamodb';
import { S3Client, GetObjectCommand, PutObjectCommand } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';

// Initialize clients
const dynamoClient = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(dynamoClient);
const s3Client = new S3Client({});

const tableName = process.env.ASSETS_TABLE_NAME!;
const bucketName = process.env.ASSETS_BUCKET_NAME!;

interface Asset {
  assetId: string;
  fileName: string;
  contentType: string;
  s3Key: string;
  s3Bucket: string;
  status: string;
  createdAt: number;
  ttl: number;
}

async function handler(event: APIGatewayProxyEvent, context: Context): Promise<APIGatewayProxyResult> {
  console.log('Asset request received', { path: event.path, method: event.httpMethod });
  
  const method = event.httpMethod;
  const pathParameters = event.pathParameters || {};
  const body = event.body ? JSON.parse(event.body) : {};
  
  const assetId = pathParameters.id;
  const action = pathParameters.action || '';
  
  try {
    switch (method) {
      case 'POST':
        return await uploadAsset(body);
      case 'GET':
        if (assetId) {
          if (action === 'download') {
            return await getDownloadUrl(assetId);
          } else {
            return await getAsset(assetId);
          }
        }
        break;
    }
    
    return {
      statusCode: 405,
      body: JSON.stringify({ error: 'Method not allowed' })
    };
  } catch (error: any) {
    console.error('Handler error', error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Internal server error' })
    };
  }
}

async function uploadAsset(body: any): Promise<APIGatewayProxyResult> {
  const assetId = body.assetId;
  const fileName = body.fileName;
  const contentType = body.contentType || 'application/octet-stream';
  
  if (!assetId || !fileName) {
    return {
      statusCode: 400,
      body: JSON.stringify({ error: 'assetId and fileName are required' })
    };
  }
  
  console.log('Creating asset metadata', { assetId, fileName });
  
  try {
    const s3Key = `assets/${assetId}/${fileName}`;
    
    // Check if asset already exists (idempotency)
    const getCommand = new GetCommand({
      TableName: tableName,
      Key: { assetId }
    });
    
    const existing = await docClient.send(getCommand);
    
    if (existing.Item) {
      console.log('Asset already exists', { assetId });
      return {
        statusCode: 200,
        body: JSON.stringify({ message: 'Asset already exists', asset: existing.Item })
      };
    }
    
    const now = Math.floor(Date.now() / 1000);
    const item: Asset = {
      assetId,
      fileName,
      contentType,
      s3Key,
      s3Bucket: bucketName,
      status: 'pending',
      createdAt: now,
      ttl: now + (30 * 24 * 60 * 60) // TTL: 30 days
    };
    
    const putCommand = new PutCommand({
      TableName: tableName,
      Item: item,
      ConditionExpression: 'attribute_not_exists(assetId)'
    });
    
    await docClient.send(putCommand);
    
    // Generate presigned URL for upload
    const putObjectCommand = new PutObjectCommand({
      Bucket: bucketName,
      Key: s3Key,
      ContentType: contentType
    });
    
    const uploadUrl = await getSignedUrl(s3Client, putObjectCommand, { expiresIn: 3600 });
    
    return {
      statusCode: 201,
      body: JSON.stringify({
        message: 'Asset metadata created',
        asset: item,
        uploadUrl
      })
    };
  } catch (error: any) {
    if (error.name === 'ConditionalCheckFailedException') {
      const getCommand = new GetCommand({
        TableName: tableName,
        Key: { assetId }
      });
      const existing = await docClient.send(getCommand);
      if (existing.Item) {
        return {
          statusCode: 200,
          body: JSON.stringify({ message: 'Asset already exists', asset: existing.Item })
        };
      }
    }
    throw error;
  }
}

async function getAsset(assetId: string): Promise<APIGatewayProxyResult> {
  console.log('Getting asset', { assetId });
  
  try {
    const command = new GetCommand({
      TableName: tableName,
      Key: { assetId }
    });
    
    const response = await docClient.send(command);
    
    if (!response.Item) {
      return {
        statusCode: 404,
        body: JSON.stringify({ error: 'Asset not found' })
      };
    }
    
    return {
      statusCode: 200,
      body: JSON.stringify(response.Item)
    };
  } catch (error) {
    console.error('DynamoDB error', error);
    throw error;
  }
}

async function getDownloadUrl(assetId: string): Promise<APIGatewayProxyResult> {
  console.log('Getting download URL', { assetId });
  
  try {
    const getCommand = new GetCommand({
      TableName: tableName,
      Key: { assetId }
    });
    
    const response = await docClient.send(getCommand);
    
    if (!response.Item) {
      return {
        statusCode: 404,
        body: JSON.stringify({ error: 'Asset not found' })
      };
    }
    
    const item = response.Item as Asset;
    
    const getObjectCommand = new GetObjectCommand({
      Bucket: item.s3Bucket,
      Key: item.s3Key
    });
    
    const downloadUrl = await getSignedUrl(s3Client, getObjectCommand, { expiresIn: 3600 });
    
    return {
      statusCode: 200,
      body: JSON.stringify({
        downloadUrl,
        fileName: item.fileName,
        contentType: item.contentType
      })
    };
  } catch (error) {
    console.error('S3/DynamoDB error', error);
    throw error;
  }
}

export { handler };

